Firestore/core/src/bundle/bundle_serializer.cc

```diff
@@ -699,37 +715,25 @@ FieldValue BundleSerializer::DecodeMapValue(JsonReader& reader,
     return {};
   }
 
-  immutable::SortedMap<std::string, FieldValue> field_values;
-  for (auto it = fields.begin(); it != fields.end(); ++it) {
-    field_values =
-        field_values.insert(it.key(), DecodeValue(reader, it.value()));
-  }
-
-  return FieldValue::FromMap(std::move(field_values));
+  google_firestore_v1_MapValue map_value{};
+  SetRepeatedField(&map_value.fields, &map_value.fields_count, fields,
+                   [&](const std::pair<std::string, json>& entry) {
+                     google_firestore_v1_MapValue_FieldsEntry result{};
+                     result.key = nanopb::MakeBytesArray(entry.first);
+                     result.value = DecodeValue(reader, entry.second);
+                     return result;
+                   });
+  return map_value;
 }
 
-FieldValue BundleSerializer::DecodeArrayValue(JsonReader& reader,
-                                              const json& array_json) const {
+google_firestore_v1_ArrayValue BundleSerializer::DecodeArrayValue(
+    JsonReader& reader, const json& array_json) const {
   const auto& values = reader.RequiredArray("values", array_json);
-  std::vector<FieldValue> field_values;
-  for (const json& json_value : values) {
-    field_values.push_back(DecodeValue(reader, json_value));
-  }
-  if (!reader.ok()) {
-    return {};
-  }
-
-  return FieldValue::FromArray(std::move(field_values));
-}
-
-FieldValue BundleSerializer::DecodeReferenceValue(
-    JsonReader& reader, const std::string& ref_string) const {
-  // Check if ref_string is indeed a valid string passed in.
-  if (!reader.ok()) {
-    return {};
-  }
 
-  return rpc_serializer_.DecodeReference(&reader, ref_string);
+  google_firestore_v1_ArrayValue array_value{};
+  SetRepeatedField(&array_value.values, &array_value.values_count, values,
```

Comment:
It not as easy anymore to do the validation since I am generally just passing through the protos. I could handle bundles differently, but for now I prefer to keep the current handling (and show a warning in DocumentSnapshot.data())